{{cookiecutter.project_name}}/.github/workflows/test.yaml

```diff
@@ -18,44 +18,61 @@ defaults:
     shell: bash -euo pipefail {0}
 
 jobs:
+  get-environments:
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.get-envs.outputs.matrix }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          filter: blob:none
+          fetch-depth: 0
+      - name: Install uv
+        uses: astral-sh/setup-uv@v5
+      - name: Get test environments
+        id: get-envs
+        run: |
+          MATRIX_JSON=$(uvx hatch env show --json | jq -c 'to_entries
+            | map(
+                select(.key | startswith("hatch-test"))
+                | {
+                    env_name: .key,
+                    name: (if (.key | contains("pre")) then .key + " (PRE-RELEASE DEPENDENCIES)" else .key end),
+                    python: .value.python
+                  }
+              )')
+          echo "matrix=${MATRIX_JSON}" | tee $GITHUB_OUTPUT
+
   test:
+    needs: get-environments
     runs-on: ${{ matrix.os }}
 
     strategy:
       fail-fast: false
       matrix:
-        include:
-          - os: ubuntu-latest
-            python: "3.10"
-          - os: ubuntu-latest
-            python: "3.12"
-          - os: ubuntu-latest
-            python: "3.12"
-            pip-flags: "--pre"
-            name: PRE-RELEASE DEPENDENCIES
-
-    name: ${{ matrix.name }} Python ${{ matrix.python }}
-
-    env:
-      OS: ${{ matrix.os }}
-      PYTHON: ${{ matrix.python }}
+        os: [ubuntu-latest]
+        include: ${{ fromJson(needs.get-environments.outputs.matrix) }}
+
+    name: ${{ matrix.name }}
 
     steps:
       - uses: actions/checkout@v4
-        with:
-          filter: blob:none
-          fetch-depth: 0
       - name: Install uv
         uses: astral-sh/setup-uv@v5
         with:
           cache-dependency-glob: pyproject.toml
+      - name: create hatch environment
+        run: |
+          uvx hatch python install ${{ matrix.python }}
+          uvx hatch env create ${{ matrix.env_name }}
```

Comment:
Apparently this requires to setup the correct python version before, while directly using `hatch test` does not. 
Additionally, it still fails with 

```
 Installing 3.8
Installed 3.8 @ /home/runner/.local/share/hatch/pythons/3.8

The following directory has been added to your PATH (pending a shell restart):

/home/runner/.local/share/hatch/pythons/3.8/python/bin
Environment `hatch-test.py3.8` is incompatible: cannot locate Python: 3.8
```

which I don't get.

Reply:
Another option is 

```yaml
- uses: astral-sh/setup-uv@v5
  with:
    python-version: ${{ matrix.python-version }}
```

Reply:
> pending a shell restart

that's it probably. 

But passing it to `setup-uv` is probably the best way forward as you suggested.